spec/libraries/course/conditional/conditional_dependency_graph_spec.rb

```diff
@@ -0,0 +1,229 @@
+require 'rails_helper'
+
+RSpec.describe Course::Conditional::ConditionalDependencyGraph do
+  class DummyConditionalCondition < ActiveRecord::Base
+    acts_as_conditional
+    acts_as_condition
+    attr_accessor :conditions
+    attr_accessor :id
+
+    def specific_conditions
+      @conditions
+    end
+
+    def dependent_object
+      self
+    end
+
+    def inspect
+      "<#{@id}>"
+    end
+
+    def self.columns
+      []
+    end
+
+    def self.build(conditions, id)
+      dummy = new
+      dummy.conditions = conditions
+      dummy.id = id
+      dummy
+    end
+  end
+
+  def create_simple_graph
+    #
+    # A ---------
+    #    |   |   |
+    #    v   v   |
+    #    C-->D   |
+    #        |   |
+    #        ->E<-
+    # B -------^
+    #
+    @a = DummyConditionalCondition.build([], 'A')
+    @b = DummyConditionalCondition.build([], 'B')
+    @c = DummyConditionalCondition.build([@a], 'C')
+    @d = DummyConditionalCondition.build([@a, @c], 'D')
+    @e = DummyConditionalCondition.build([@a, @b, @d], 'E')
+
+    [@c, @a, @b, @e, @d]
+  end
+
+  def create_cyclic_graph
+    #
+    # A ----> B ----> C
+    # ^               |
+    # |               |
+    #  ----------------
+    #
+    @a = DummyConditionalCondition.build([], 'A')
+    @b = DummyConditionalCondition.build([@a], 'B')
+    @c = DummyConditionalCondition.build([@b], 'C')
+    @a.conditions.append(@c)
+
+    [@b, @a, @c]
+  end
+
+  def create_disconnected_graph
+    #
+    # A ----> B ----> C
+    #
+    # D ----> E ----> F
+    #
+    # G ----> H ----> I
+    #
+    @a = DummyConditionalCondition.build([], 'A')
+    @b = DummyConditionalCondition.build([@a], 'B')
+    @c = DummyConditionalCondition.build([@b], 'C')
+
+    @d = DummyConditionalCondition.build([], 'D')
+    @e = DummyConditionalCondition.build([@d], 'E')
+    @f = DummyConditionalCondition.build([@e], 'F')
+
+    @g = DummyConditionalCondition.build([], 'G')
+    @h = DummyConditionalCondition.build([@g], 'H')
+    @i = DummyConditionalCondition.build([@h], 'I')
+
+    [@b, @d, @e, @h, @g, @a, @i, @c, @f]
+  end
+
+  describe '.build' do
+    subject { Course::Conditional::ConditionalDependencyGraph }
+
+    context 'when dependencies do not contain cycle' do
+      before(:each) do
+        @simple_graph = create_simple_graph
+      end
+
+      it 'build dependency graph' do
+        expect(subject.build(@simple_graph)).
+          to be_instance_of(Course::Conditional::ConditionalDependencyGraph)
+      end
+    end
+
+    context 'when dependencies contains cycle' do
+      before(:each) do
+        @cyclic_graph = create_cyclic_graph
+      end
+
+      it 'raise ArgumentError' do
+        expect { subject.build(@cyclic_graph) }.to raise_error(ArgumentError)
+      end
+    end
+  end
+
+  describe '#resolve' do
+    context 'simple graph' do
+      before(:each) do
+        @simple_graph = create_simple_graph
+        allow_any_instance_of(DummyConditionalCondition).
```

Comment:
What you can do is you want only these elements to be affected, you should .each the nodes and then allow each of them. in future when we do run specs in parallel, this would be less of a problem.